lib/nghttp2_session.c

```diff
@@ -4410,6 +4410,12 @@ int nghttp2_session_on_settings_received(nghttp2_session *session,
     return session_call_on_frame_received(session, frame);
   }
 
+  if (!session->remote_settings_received) {
+    session->remote_settings.max_concurrent_streams =
+        NGHTTP2_DEFAULT_MAX_CONCURRENT_STREAMS;
+    session->remote_settings_received = 1;
+  }
```

Comment:
I'm wondering setting max_concurrent_streams to max somewhere near https://github.com/nghttp2/nghttp2/blob/bfd4d8b052a594eb61b39d0ff260fbd19856763d/lib/nghttp2_session.c#L5421 might make the PR simpler because we don't have to add the new field to nghttp2_session.

Reply:
My suggestion doesn't work when we see the some part of SETTINGS and before calling nghttp2_session_send.  I think your original PR looks good to me.

Reply:
Yeah, I think that removing initial limit in `NGHTTP2_IB_READ_FIRST_SETTINGS` could open the floodgates in case of partial read of the `SETTINGS` frame from a peer that advertises `SETTINGS_MAX_CONCURRENT_STREAMS`.

Thanks for investigating!